Exercises 2, 3, 8

lupreCSC opened the pull request:
Instructions, scripts and reference solutions for exercises
- [02_Using_the_LUMI_web_interface](https://github.com/Lumi-supercomputer/Getting_Started_with_AI_workshop/tree/main/02_Using_the_LUMI_web_interface)
- [03_Your_first_AI_training_job_on_LUMI](https://github.com/Lumi-supercomputer/Getting_Started_with_AI_workshop/tree/main/03_Your_first_AI_training_job_on_LUMI)
- [08_Scaling_to_multiple_GPUs](https://github.com/Lumi-supercomputer/Getting_Started_with_AI_workshop/tree/main/08_Scaling_to_multiple_GPUs)

These are mostly feature-complete but will still be polished a bit.
[xx_Exercise_base](https://github.com/Lumi-supercomputer/Getting_Started_with_AI_workshop/tree/csc_exercise_dev/xx_Exercise_base) will also be removed before merge.

lupreCSC committed:
Adapted imdb example from PDL course. [work in progress]

Bit less evolved set up for the dataset and no concatenation of training samples.

Runs 5000 steps with batch size 16 on single GPU in ~10 minutes.
Tried to adapt to multigpu as well, but not getting it to run properly just yet. Refuses to run without torchrun. With torchrun it runs but takes the same amount of time, so I suspect it just spawn two independent
processes that don't talk to each other. Needs fixing.

Uses CSC pytorch module so far, will need to adjust.

lupreCSC committed:
Adjusted example

- fixed some small potential issues with transformers
- improved multigpu/-node configuration setup
- added run scrip with CPU bindings
- switched to different, larger pretrained model

lupreCSC committed:
removed custom binary path after singularity_wrapper fix

lupreCSC committed:
keep dataset in memory

lupreCSC committed:
fix: multigpu script now loads correct device for final prediction

lupreCSC committed:
added scripts for creating and running with LUST easybuild environment

lupreCSC committed:
fixing reference comment to LLM

lupreCSC committed:
Moved exercise prototype into temporary subdirectory

lupreCSC committed:
Merge branch 'main' into exercise_dev

lupreCSC committed:
Added exercise description and notebook for 02_Using_the_LUMI_web_interface.

lupreCSC committed:
Small fixes for exercises of 02_Using_the_LUMI_web_interface

lupreCSC committed:
Initial exercise setup for 03_Your_first_AI_training_job_on_LUMI

lupreCSC committed:
added initial exercise instructions for 08_Scaling_to_multiple_GPUs

lupreCSC committed:
Make the actual task in exercises for section 3 and 8 more explicit.

lupreCSC committed:
Added reference solution script for 03_Your_first_AI_training_job_on_LUMI

lupreCSC committed:
Completed reference solution for 03_Your_first_AI_training_job_on_LUMI

lupreCSC committed:
Updates to exercises 3 and 8

- added reference solution for exercise 8 (as of yet untested)
- added software environment setup to run.sh templates
- wrapped main code in python scripts in `if __name__ == '__main__'`

lupreCSC committed:
exercise 8 reference solution now prints the cpus it's binding to

lupreCSC committed:
Fixing reference solution for exercise 8.

Some bugs in python script and slurm batch file.
Also added timing of model loading and moved
HF cache to flash to increase model loading speed.

lupreCSC committed:
Adjusting all remaining scripts for exercises 3 and 8 to changes in previous commit

- using flash storage
- timing model loading

lupreCSC committed:
Added mention of LUMI web interface editor in exercise 3 and example links to bonus materials for exc 3 and 8

lupreCSC committed:
Added container recipe for exercises 3 and 8 in bonus_material/

lupreCSC committed:
fix: changed environment name in pytorch_transformers container

lupreCSC committed:
Small fixes to exercises 2, 3, 8

- renamed all index.md to README.md
- Exercise 2 instructions no longer mention copying the check out repo, but have a note about using Jupyter lab UI for cloning git
-  all exercise instructions mention use of course reservations
- full node batch scripts in reference solutions no longer use --mem=0
- added note about future availability of singularity-userfilesystems and singularity-CPEbits modules
- Exercise 3 mentions required command line arguments for python training script
- Exercise 8 reference solutiosn have note about CPU bind masks only being useful for full node allocations

lupreCSC committed:
Added print of device name in Exercise 2 notebook

Chroxvi commented:
@lupreCSC Is this ready for review and merging?

lupreCSC commented:
Doing some small changes at the moment, but will have those ready later today

lupreCSC committed:
Exercises 2,3,8: Fix alert boxes not rendering on github

.. despite the syntax being as described in the github docs: https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#alerts

Now changed to normal quote boxes with bolded headlines

lupreCSC committed:
Exercise 8: Less convoluated cpu binds from Python code

Previous approach was a bit too complex, now simpler and more in line with the additional examples in the bonus material.

lupreCSC committed:
Exercises 3,8: improved formatting of python training script

using `black`

lupreCSC committed:
Exercise 8: removed incomplete comment

lupreCSC committed:
Added exercise_container_reciptes to bonus materials index.md

lupreCSC committed:
Added GPT-neo finetuning multinode example to bonus material

currently untested

lupreCSC committed:
Exercise 3, 8: Various small fixes to changes in scripts.

lupreCSC committed:
Removed xx_Exercise_base folder

has been fully integrated into Exercises 2,3,8 as well as bonus material

Chroxvi committed:
Updated pytorch_transformers.yml to include ray-tune.

Chroxvi commented on 02_Using_the_LUMI_web_interface/GPT-neo-IMDB-introduction.ipynb:
```diff
@@ -0,0 +1,191 @@
+    "import torch\n",
+    "import os\n",
+    "from datasets import load_dataset\n",
+    "from transformers import AutoModelForCausalLM, AutoTokenizer"
```
I really like PEP8, including it's [style guide on ordering of imports](https://peps.python.org/pep-0008/#imports).

```suggestion
    "import os\n",

    "import torch\n",
    "from datasets import load_dataset\n",
    "from transformers import AutoModelForCausalLM, AutoTokenizer"
```

lupreCSC replied:
done, also for the python scripts in the other exercises

Chroxvi commented on 02_Using_the_LUMI_web_interface/README.md:
```diff
@@ -0,0 +1,42 @@
+
+    In this exercise you will gain first experience with using the LUMI web interface to navigate files and directories on the LUMI supercomputer. You will also set up your own copy of the exercise repository on the system, so that you can work on them without interfering with the other course participants.
+   
+   1. Log in to the LUMI web interface: https://lumi.csc.fi
```
I believe in most browsers, you need to either go to `lumi.csc.fi` or prepend both `https://` and `www.` to have it resolve correctly.
```suggestion
   1. Log in to the LUMI web interface: https://www.lumi.csc.fi
```

lupreCSC replied:
done

Chroxvi commented on 02_Using_the_LUMI_web_interface/README.md:
```diff
@@ -0,0 +1,42 @@
+   2. Create your own scratch subdirectory in `/scratch/project_465001063/`. Use your username for the directory name. You can either
+        - Use the built-in file explorer, or
+        - Use a login node shell.
+   3. Clone the exercise repository to your scratch folder `/scratch/project_465001063/<username>`. You can either
```
Should we link the GH repo here or is it mentioned somewhere else?
```suggestion
   3. Clone the [exercise repository](https://github.com/Lumi-supercomputer/Getting_Started_with_AI_workshop) to your scratch folder `/scratch/project_465001063/<username>`. You can either
```

lupreCSC replied:
done

Chroxvi commented on 02_Using_the_LUMI_web_interface/README.md:
```diff
@@ -0,0 +1,42 @@
+        - start a Jupyter lab job and use the Jupyter lab UI for cloning Git repositories (check [this blog post](https://blog.reviewnb.com/jupyterlab-git-extension/) for a visual guide for how it works). See point 2.1 below for instructions on how to start the job from the LUMI web interface.
+   4. Get familiar with the exercise repository layout.
+
+2. Start a an interactive Jupyter lab job and run inference with GPT-neo.
```
```suggestion
2. Start an interactive Jupyter lab job and run inference with GPT-neo.
```

lupreCSC replied:
fxed

Chroxvi commented on 02_Using_the_LUMI_web_interface/README.md:
```diff
@@ -0,0 +1,42 @@
+    1. Start an interactive Jupyter session: Open the Jupyter app in the LUMI webinterface and set the following settings before pressing `Launch`
+        - Project: `project_465001063`
+        - Reservation: Use the course reservation (there should only be one option)
+        - Partition: `dev-g` or `small-g`
```
If our reservation is for `small-g`, this needs to only say `small-g`, right?

Chroxvi commented on 03_Your_first_AI_training_job_on_LUMI/GPT-neo-IMDB-finetuning.py:
```diff
@@ -0,0 +1,244 @@
+    # Then we determine the device on which to train the model.
+    print("Using PyTorch version:", torch.__version__)
+    if torch.cuda.is_available():
+        device = torch.device("cuda")
```
I believe when using the `transformers.Trainer`, you don't need to worry about managing devices yourself, including transferring the model and dataset (batches) to the device. This is automagically handled by the `Trainer` and its `Accelerator` based on the `LOCAL_RANK` env var set by torchrun (or manually). You may want to consider leaving out this manual device handling.

lupreCSC replied:
You are right. I was thinking of having it there so that people learn how to do it anyways, but it might raise some questions on why we are doing it here...

Chroxvi replied:
That's a fair point. Maybe you can also just mention it at some point during the workshop.

Chroxvi commented on 03_Your_first_AI_training_job_on_LUMI/GPT-neo-IMDB-finetuning.py:
```diff
@@ -0,0 +1,244 @@
+        # Let's print a few sample generated reviews; this is the same as in the previous exercise
+        # but now we use the finetuned model
+        prompt = "The movie 'How to run ML on LUMI - A documentation' was great because"
+        inputs = tokenizer(prompt, return_tensors="pt").to(device)
```
If removing the manual device handling for the `Trainer`, you still need to transfer the input here. In that case, `trainer.accelerator.device` gives you a reference to the relevant device.

Chroxvi commented on 03_Your_first_AI_training_job_on_LUMI/README.md:
```diff
@@ -0,0 +1,84 @@
+
+## Examples
+
+[comment]: <> (List your examples from the lecture here and provide the necessary links to scripts, notebooks, etc. to run them on LUMI)
```
```suggestion
```

lupreCSC replied:
done

Chroxvi commented on 03_Your_first_AI_training_job_on_LUMI/README.md:
```diff
@@ -0,0 +1,84 @@
+
+1. Familiarise yourself with the training script.
+   
+   **Task**: Check the training script [03_Your_first_AI_training_job_on_LUMI/GPT-neo-IMDB-finetuning.py](03_Your_first_AI_training_job_on_LUMI/GPT-neo-IMDB-finetuning.py). 
```
When listing the exercises in the README.md for lecture 2, you gave a short introduction to the goal of the exercise and used a numbered list to highlight the task (instead of **Task**: as used here). If you provide a consistent way of listing the exercises in the README files in this PR, I will adjust the README files for the other lectures accordingly.

lupreCSC replied:
Done. They now all use the numbering as in Ex 2.

Chroxvi commented on 03_Your_first_AI_training_job_on_LUMI/reference_solution/run.sh:
```diff
@@ -0,0 +1,41 @@
+
+# Set up the software environment
+# NOTE: these modules will be available from the LUMI system stack after July 2024 and the "module use" line will no longer work
+module purge
```
Note the remark about using `module purge` with the sticky software stack modules on LUMI in the [LUMI Docs](https://docs.lumi-supercomputer.eu/runjobs/lumi_env/Lmod_modules/#loading-and-unloading-modules). I don't know if `module --force purge` is a good or bad idea here.

lupreCSC replied:
I think we don't want to `--force` here.

Chroxvi commented on 03_Your_first_AI_training_job_on_LUMI/reference_solution/run.sh:
```diff
@@ -0,0 +1,41 @@
+# Set up the software environment
+# NOTE: these modules will be available from the LUMI system stack after July 2024 and the "module use" line will no longer work
+module purge
+module use /project/project_465001063/modules
```
Note that @Danzelot is working on moving these modules to `/appl/local/training/modules/AI-20240529/` such that they remain after project 465001063 has been deleted.

lupreCSC replied:
I'll update when that's done.

Chroxvi commented on 03_Your_first_AI_training_job_on_LUMI/reference_solution/run.sh:
```diff
@@ -0,0 +1,41 @@
+#SBATCH --cpus-per-task=7
+#SBATCH --mem-per-gpu=60G
+#SBATCH --time=0:15:00
+
```
Should this include a `--reservation` specification?

lupreCSC replied:
It probably should, but I don't know which reservation to use, so I was thinking of adding it later.

lupreCSC replied:
done

Chroxvi commented on 08_Scaling_to_multiple_GPUs/README.md:
```diff
@@ -0,0 +1,83 @@
+   - `MASTER_ADDR`: The FQDN of the host that is running worker with rank 0; used to initialize the Torch Distributed backend.
+   - `MASTER_PORT`: The port on the `MASTER_ADDR` to use.
+
+   Using torchrun, every process sees all GCDs on a node, so we need to make sure that our script selects the GCD it is going to train on according to its local rank.
```
See my previous comment about the magic of `Trainer` and its `Accelerator`.

Chroxvi commented on 08_Scaling_to_multiple_GPUs/README.md:
```diff
@@ -0,0 +1,83 @@
+   However, we are using the HuggingFace Trainer which will automatically take care of setting up data-parallel training for the model if it detects the environment variables set by torchrun, so we do not need to handle that manually.
+
+   **Task**: You will need to make the following changes to the script:
+   - select the correct PyTorch device
```
Strictly, not needed, I think, if you use the magic of the `Trainer`.

Chroxvi commented on 08_Scaling_to_multiple_GPUs/README.md:
```diff
@@ -0,0 +1,83 @@
+    - the correct slurm partition
+    - number of GPUs requested (8)
+    - number of CPUs requested
+    - RAM requested (you can use `0` to reserve all memory on the node)
```
We generally recommend requsting `480G` memory instead of `0` when using a full node. This reserves some memory for the system processes and any memory leaks from previous jobs, which we have had tickets about in the past when setting `0`.

lupreCSC replied:
done.

Chroxvi commented on 08_Scaling_to_multiple_GPUs/README.md:
````diff
@@ -0,0 +1,83 @@
+   >
+   > You can get the hostname of the node running the rank 0 process using the command
+   > ```
+   > scontrol show hostname ${SLURM_NODELIST} | head -n 1
````
I think a simple `hostname`, maybe `hostname -i`, is sufficient here (as you suggested in https://docs.csc.fi/support/tutorials/ml-multi/). It worked in my examples at least...

Chroxvi commented on 08_Scaling_to_multiple_GPUs/reference_solution/run_torchrun.sh:
```diff
@@ -0,0 +1,50 @@
+export OUTPUT_DIR=$SCRATCH/$USER/data/
+export LOGGING_DIR=$SCRATCH/$USER/runs/
+export MODEL_NAME=gpt-imdb-model-${SLURM_JOBID}
+
```
I am not entirely sure, but I think you also need to set `MIOPEN_USER_DB_PATH` and `MIOPEN_CUSTOM_CACHE_DIR` to some /tmp based locatgion when using multiple processes to avoid a potential race condition problem when the MIOpen DB/Cache is on the Lustre file system.

lupreCSC replied:
As far as I know that one is only really used for convolutional kernels, which we don't use in this example. At least I have never encountered problems with it here

lupreCSC replied:
but I can add them if you think that would be better

Chroxvi replied:
"Less is more", so don't include it unless it's necessary. I think we should keeps things as simple as possible. I just wasn't aware that this was only needed when certain MIOpen kernels are used.

lupreCSC replied:
Maybe we can confirm this today with Samuel

Chroxvi commented on bonus_material/GPT_neo_IMDB_multinode/GPT-neo-IMDB-finetuning.py:
```diff
@@ -0,0 +1,302 @@
+#!/usr/bin/env python
```
I guess this is meant as a helping hand to Samuel for the lecture 10 exercises? At least I believe that is should be part of the lecture 10 exercises and not included as bonus material.

Chroxvi commented on bonus_material/GPT_neo_IMDB_multinode/run_no_torchrun.sh:
```diff
@@ -0,0 +1,59 @@
+export MASTER_PORT=25900
+export WORLD_SIZE=$SLURM_NPROCS
+export LOCAL_WORLD_SIZE=$WORLD_SIZE
+
```
I believe you need to set `export NCCL_SOCKET_IFNAME=hsn` here as well when using the `aws-ofi-rccl` plugin.

lupreCSC replied:
So the container comes with the plugin but doesn't set the variable already?

Chroxvi replied:
I don't think so. It seems that @sfantao only sets it [when testing the containers](https://github.com/search?q=repo%3Asfantao%2Flumi-containers%20NCCL_SOCKET_IFNAME&type=code). It's an interesting idea though, so I will add it to the agenda for the next LUMI container group meeting to discuss if this should be set in containers that have the aws-ofi-rccl plugin installed. The [CXI_* variables are already set that way](https://github.com/sfantao/lumi-containers/blob/aeeb8993208ac05ed2294eaf44b31cf0f7be88a8/common/Dockerfile.aws-ofi-rccl#L25).

Chroxvi replied:
But if you can run this without setting `NCCL_SOCKET_IFNAME`, it makes me wonder when that environment variable is really needed...

lupreCSC replied:
Iirc, not setting this will still allow a multi-node run to work but it might use tcp instead of slingshot, so poorer performance results. I'll set it but I think it should be set in the container because it's as much part of proper network setup as the `aws-ofi-rccl` plugin.

lupreCSC replied:
done

Chroxvi reviewed (changes requested):
I think this is really nice!

I have added some comments, that you may want to consider. If you like your solution better than my suggestion, please keep your solution.

lupreCSC committed:
Exercise 2,3,8: PEP8 style compliant import ordering

lupreCSC committed:
Exercise 3,5,8: Various small improvements from PR review

lupreCSC committed:
Exercise 3,8 now use the same way of numbering exercise tasks as Ex 2

lupreCSC committed:
GPT IMDb multinode bonus: setting NCCL_SOCKET_IFNAME=hsn

lupreCSC committed:
Exercises 2,3,8: added course reservation

lupreCSC committed:
Ex 3,8: updated paths to course modules (module use)

lupreCSC committed:
Ex 3,8: commented reservation line in reference solution

so that they can be run always and the line can be uncommented during the course to use the reservation

lupreCSC committed:
explicit LUMI module version in bonus_material/exercise_container_recipes/build_pytorch_transformers.sh

Chroxvi commented:
@lupreCSC Do you consider this ready to be merged?

lupreCSC commented:
From my side yes, unless you want me to get rid of the manual device loading

Chroxvi reviewed (approved):
All looks good to me now. I'll merge
